src/node_api.cc

```diff
@@ -3553,3 +3554,250 @@ napi_status napi_run_script(napi_env env,
+      return;
+    }
+    ts_fn->idle_running = true;
+  }
+}
+
+static void napi_threadsafe_function_default_call_js(napi_env env,
+                                                     napi_value js_cb,
+                                                     void* hint,
+                                                     void* data) {
+  napi_value recv;
+  napi_status status;
+  status = napi_get_global(env, &recv);
+  if (status != napi_ok) {
+    napi_throw_error(env, nullptr,
+        "N-API thread-safe function: failed to retrieve global object");
+    return;
+  }
+
+  status = napi_call_function(env, recv, js_cb, 0, nullptr, nullptr);
+  if (status != napi_ok && status != napi_pending_exception) {
+    napi_throw_error(env, nullptr,
+        "N-API thread-safe function: failed to call JS callback");
+    return;
+  }
+}
+
+NAPI_EXTERN napi_status
+napi_create_threadsafe_function(napi_env env,
+                                napi_value func,
+                                void* hint,
+                                size_t max_queue_size,
+                                napi_threadsafe_function_call_js call_js_cb,
+                                napi_threadsafe_function* result) {
+  CHECK_ENV(env);
+  CHECK_ARG(env, func);
+  CHECK_ARG(env, result);
+
+  napi_threadsafe_function ts_fn = nullptr;
+  napi_valuetype func_type;
+  napi_status status = napi_typeof(env, func, &func_type);
+  if (status != napi_ok) {
+    return status;
+  }
+
+  uv_loop_t* loop = uv_default_loop();
+
+  status = napi_generic_failure;
+
+  if (func_type != napi_function) {
+    goto failure;
+  }
+
+  ts_fn = new napi_threadsafe_function__;
+  if (ts_fn == nullptr) {
+    goto failure;
+  }
+
+  if (uv_async_init(loop, &ts_fn->async, napi_threadsafe_function_cb) != 0) {
+    delete ts_fn;
+    goto failure;
+  }
+
+  if (uv_mutex_init(&ts_fn->mutex) != 0) {
+    goto close_async;
+  }
+
+  status = napi_create_reference(env, func, 1, &ts_fn->ref);
+  if (status != napi_ok) {
+    goto destroy_mutex;
+  }
+
+  if (max_queue_size > 0) {
+    if (uv_cond_init(&ts_fn->cond) != 0) {
+      goto destroy_reference;
+    }
+  }
+
+  if (uv_idle_init(loop, &ts_fn->idle) != 0) {
+    goto destroy_cond;
+  }
+
+  ts_fn->call_js_cb = call_js_cb == nullptr ?
+      napi_threadsafe_function_default_call_js : call_js_cb;
+  ts_fn->hint = hint;
+  ts_fn->env = env;
+  ts_fn->max_queue_size = max_queue_size;
+  ts_fn->idle_running = false;
+
+  *result = reinterpret_cast<napi_threadsafe_function>(ts_fn);
+  return napi_clear_last_error(env);
+
+destroy_cond:
+  if (max_queue_size > 0) {
+    uv_cond_destroy(&ts_fn->cond);
+  }
+destroy_reference:
+  status = napi_delete_reference(env, ts_fn->ref);
+
+  // This is a failure path, so even if we successfully destroy the reference,
+  // we have still failed to create the thread-safe function.
+  if (status == napi_ok) {
+    status = napi_generic_failure;
+  }
+destroy_mutex:
+  uv_mutex_destroy(&ts_fn->mutex);
```

Comment:
Can we use the helpers from `node_mutex.h` for these things? They really improve readability for C++, all these `goto`s are really hard to wrap your head around…

Reply:
Those abort if the mutex is not created succesfully. N-API returns `napi_generic_failure`.